src/main/java/appeng/recipes/factories/recipes/PartRecipeFactory.java

```diff
@@ -45,25 +60,149 @@ public IRecipe parse( JsonContext context, JsonObject json )
+		JsonArray patternJ = JsonUtils.getJsonArray( json, "pattern" );
+
+		if( patternJ.size() == 0 )
+		{
+			throw new JsonSyntaxException( "Invalid pattern: empty pattern not allowed" );
+		}
+
+		String[] pattern = new String[patternJ.size()];
+		for( int x = 0; x < pattern.length; ++x )
+		{
+			String line = JsonUtils.getString( patternJ.get( x ), "pattern[" + x + "]" );
+			if( x > 0 && pattern[0].length() != line.length() )
+			{
+				throw new JsonSyntaxException( "Invalid pattern: each row must  be the same width" );
+			}
+			pattern[x] = line;
+		}
+
+		CraftingHelper.ShapedPrimer primer = new CraftingHelper.ShapedPrimer();
+		primer.width = pattern[0].length();
+		primer.height = pattern.length;
+		primer.mirrored = JsonUtils.getBoolean( json, "mirrored", true );
+		primer.input = NonNullList.withSize( primer.width * primer.height, net.minecraft.item.crafting.Ingredient.EMPTY );
+
+		Set<Character> keys = Sets.newHashSet( ingMap.keySet() );
+		keys.remove( ' ' );
+
+		int x = 0;
+		for( String line : pattern )
+		{
+			for( char chr : line.toCharArray() )
+			{
+				net.minecraft.item.crafting.Ingredient ing = ingMap.get( chr );
+				if( ing == null )
+				{
+					throw new JsonSyntaxException( "Pattern references symbol '" + chr + "' but it's not defined in the key" );
+				}
+				primer.input.set( x++, ing );
+				keys.remove( chr );
+			}
+		}
+
+		if( !keys.isEmpty() )
+		{
+			throw new JsonSyntaxException( "Key defines symbols that aren't used in pattern: " + keys );
+		}
+
+		JsonObject resultObject = (JsonObject) json.get( "result" );
+		int count = JsonUtils.getInt( resultObject, "count", 1 );
+
+		String ingredient = resultObject.get( "part" ).getAsString();
+		Object result = (Object) Api.INSTANCE.registries().recipes().resolveItem( AppEng.MOD_ID, ingredient );
+		if( result instanceof ResolverResult )
+		{
+			ResolverResult resolverResult = (ResolverResult) result;
+
+			Item item = Item.getByNameOrId( AppEng.MOD_ID + ":" + resolverResult.itemName );
+			if( item == null )
+			{
+				AELog.warn( "item was null for " + resolverResult.itemName + " ( " + ingredient + " )!" );
+				throw new JsonSyntaxException( "Got a null item for " + resolverResult.itemName + " ( " + ingredient + " ). This should never happen!" );
+			}
+			ItemStack itemStack = new ItemStack( item, count, resolverResult.damageValue, resolverResult.compound );
+
+			return new ShapedOreRecipe( new ResourceLocation( group ), itemStack, primer );
+		}
+
+		// Should never reach this part unless mangled JSON or bug in AE.
+		throw new JsonSyntaxException( "Couldn't find the resulting item in AE. This means AE was provided a recipe that it shouldn't be handling.\nWas looking for : '" + ingredient + "'." );
+	}
+
+	// Copied from ShapelessOreRecipe.java, modified a bit.
+	private static ShapelessOreRecipe shapelessFactory( JsonContext context, JsonObject json )
+	{
+		String group = JsonUtils.getString( json, "group", AppEng.MOD_ID + ":part_shapeless_crafting" );
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Move the declaration of "group" closer to the code that uses it. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1941)